bin/propolis-server/src/lib/initializer.rs

```diff
@@ -586,22 +593,22 @@ impl<'a> MachineInitializer<'a> {
             Nvme,
         }
 
-        for (disk_name, disk) in &self.spec.disks {
+        for (device_id, disk) in &self.spec.disks {
```

Comment:
i do think that `disk_name` was a good name for this one - it's a `SpecKey`, and it _could_ be a `Uuid` if something sent one, but Nexus currently sends a `Name` and changing it would (for better or worse) change `/dev/disk/by-id/` paths guest VMs should be able to depend on.

i recognize that you're not changing `disk_name` to something else (yet?), but i happen to _also_ think we should change this _name_ to an ID. if you agree, it'd be nice to change the name of this variable when we do it :grin:

Reply:
Great catch here, thanks! I *think* the disk serials will remain stable over migrations, at least, but you're right, without some care this will change the serial number strings we use for NVMe devices across a reboot.

One of my goals for this layer is to divorce components' keys from their behaviors--i.e. two devices with the same component specs should have the same guest-visible behavior even if they had different keys. In this particular case, we can achieve that by adding a `serial_number` field to the `NvmeDisk` component, and then populating it with whatever makes sense in the control plane (the actual disk name, I believe). If that sounds good, I can give this a try when I next make updates to this change.

Reply:
oh yes, love that idea!

Reply:
Fixed in 6e82d49.

I represented the `serial_number` as a `[u8; 20]` instead of a `String` to try to make clients be explicit about padding. Just taking a `String` leaves Propolis to decide whether to pad with 0s, spaces, or something else; this method prevents us from accidentally changing padding disciplines across a migration. Another approach would be to have a `serial_number: String` and a `serial_padding_byte: u8` in the spec itself; this is more verbose, but it's convenient enough for clients that I added a support function to `propolis-client` to generate serial number byte arrays from a `str` and padding byte.

I'm not strongly beholden to this approach--let me know if you think a different one makes more sense.

Reply:
(i do have thoughts, but i think on `disk_name`/id/serial we're in a good spot. thanks! i'll comment elsewhere as appropriate.)